### PR TITLE
Generate a tree for downstream documentation

### PR DESCRIPTION
**Pull request by gsmet**
This is a first version of the tool to generate a tree for the downstream doc.

It generates two things:
- the list of files
- a full tree containing all the needed files that we could push to the repository

It might not be completely perfect, we can tweak it when we add files.

Structure of the tree is the following (but can be adjusted depending on the requirements):

```
downstream-tree
├── _attributes.adoc
├── datasource.adoc
├── deploying-to-openshift.adoc
├── _generated
│   └── config
│       ├── quarkus-agroal.adoc
│       ├── quarkus-datasource.adoc
│       ├── quarkus-oidc.adoc
│       ├── quarkus-openshift-openshift-config.adoc
│       ├── quarkus-reactive-datasource.adoc
│       ├── quarkus-reactive-db2-client.adoc
│       ├── quarkus-reactive-mssql-client.adoc
│       ├── quarkus-reactive-mysql-client.adoc
│       ├── quarkus-reactive-oracle-client.adoc
│       ├── quarkus-reactive-pg-client.adoc
│       └── quarkus-vertx-http-config-group-form-auth-config.adoc
├── images
│   ├── authorization_code_flow.png
│   ├── security-architecture-overview.png
│   ├── security-bearer-token-authorization-mechanism-1.png
│   └── security-bearer-token-authorization-mechanism-2.png
├── _includes
│   ├── devtools
│   │   ├── build.adoc
│   │   ├── build-native.adoc
│   │   ├── create-app.adoc
│   │   ├── dev.adoc
│   │   └── extension-add.adoc
│   └── prerequisites.adoc
├── security-architecture.adoc
├── security-authentication-mechanisms.adoc
├── security-authorize-web-endpoints-reference.adoc
├── security-basic-authentication.adoc
├── security-basic-authentication-howto.adoc
├── security-basic-authentication-tutorial.adoc
├── security-identity-providers.adoc
├── security-jpa.adoc
├── security-oidc-bearer-token-authentication.adoc
├── security-oidc-bearer-token-authentication-tutorial.adoc
├── security-oidc-code-flow-authentication.adoc
├── security-oidc-code-flow-authentication-tutorial.adoc
├── security-oidc-configuration-properties-reference.adoc
├── security-openid-connect-client-reference.adoc
├── security-overview.adoc
├── security-proactive-authentication.adoc
├── security-vulnerability-detection.adoc
└── update-quarkus.adoc
```

### CODE REVIEW

**Comment by gsmet** [comment]
I made several improvement in a follow-up commit, including rewriting the links that are not handled downstream to links to quarkus.io.

The tree doesn't contain `_attributes-local.adoc` anymore as we don't want to override it.

The `quarkusio-guides` attribute needs to be set in `_attributes-local.adoc` to the root guides URL (for instance for `main`, `https://quarkus.io/version/main/guides/`).

Here is an example of what we obtain, please check that it works for you (I made a few sanity checks but I could have missed something):
[downstream-tree.zip](https://github.com/quarkusio/quarkus/files/12238077/downstream-tree.zip)

File: docs/downstreamdoc.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+guides:
+  - src/main/asciidoc/datasource.adoc
+  - src/main/asciidoc/deploying-to-openshift.adoc
```
Comment (michelle-purcell) [inline docs/downstreamdoc.yaml:3]: Please add one more... :pray: 

```suggestion
  - src/main/asciidoc/deploying-to-openshift.adoc
  - src/main/asciidoc/logging.adoc
```

**Comment by michelle-purcell** [comment]
@gsmet - Thank you. Super job :100:  I reviewed the generated download.zip output. It looks great. 
The archive is much smaller and therefore much easier for us to work with :+1: 

So far my local downstream doc build looks good with the zip content. I will push all the way to PV2 soon and come back if there are any issues.

 Just one thing I noticed is that the other naked anchor limitation post-fix (to convert <<section-ID>> to an xref anchor)  is not applied to the asciidoc file in the download.zip file above. For example:

image

Is that expected because this PR is in progress? Flagging just in case. Thanks.

**Comment by michelle-purcell** [comment]
Regarding this comment:

_The quarkusio-guides attribute needs to be set in _*attributes-local. adoc*_ to the root guides URL (for instance for main, https://quarkus.io/version/main/guides - no / at the end)._

This is now implemented downstream. Thanks.

**Comment by michelle-purcell** [comment]
In the downstream product docs, the first instance of Quarkus in the text must always state the full product name: *Red Hat build of Quarkus*, which means another manual edit in GitLab after import.

Would it be possible to add to the existing post-script job a script that finds the first instance of either `Quarkus` or `{project-name}` in a main guide adoc and replace it with `Red Hat build of Quarkus`? 

_Let me know if we need to move this specific request elsewhere if this is completely outside the scope of this PR/priority. Thanks._

**Comment by gsmet** [comment]
Let's not add more requirements to this PR for now, please.

**Comment by gsmet** [comment]
> Just one thing I noticed is that the other naked anchor limitation post-fix (to convert <> to an xref anchor) is not applied to the asciidoc file in the download.zip file above. For example:

Yes, this is due to me not pointing to the transformed files but to a copy of the source files, I'm fixing it and will push a new version soon.

**Comment by gsmet** [comment]
@michelle-purcell I added the Logging guide and fixed the generated source issue.

This file should be much better: 
[downstream-tree.zip](https://github.com/quarkusio/quarkus/files/12259165/downstream-tree.zip)

**Comment by michelle-purcell** [comment]
@gsmet - Thank you so much. This is great.

Is it possible to regenerate the `download-test.zip` on the `3.2` branch source?

I see some of the changes in the diff might be post-3.2 content. 

For example:

image

**Comment by gsmet** [comment]
Yes, don't use these files as is for now. What we need is to validate the structure and the behavior. We will use this script on whatever branch we need, once we validate it produces what we need.

**Review by michelle-purcell — APPROVED** [review]

**Comment by gsmet** [comment]
I merged the script. I will backport it to 3.2 so that you can use it there to generate a 3.2 downstream tree.

**Comment by michelle-purcell** [comment]
@gsmet - Awesome work here :100:  :superhero: Thanks so much :green_heart:

**Comment by inoxx03** [comment]
@michelle-purcell Awesome work on the vale tweaks:)

@gsmet many thanks for merging this PR!  

Thank you both:) 

@michelle-purcell about your request to add a feature that swaps in Red hat brand-approved product names for donwstreaming: I think @rolfedh has submitted a draft PR of a prototype of a script that automates this change: https://github.com/quarkusio/quarkus/pull/35264

I also agree with @gsmet that we should add that capability in a follow-up PR:)

**Comment by michelle-purcell** [comment]
@gsmet - I ran the latest Maven doc build on `main` but I cannot find the download.zip (name??) archive output in the target folder. Which folder should it get created in? Thanks.

**Pull request merged**